### ISSUE TITLE
9.2.3 - Crash When Attempting to "Star" Professions

### ISSUE THREAD

**Issue by MystLeissa** [StrangeLoopGames/EcoIssues#20307]
Apparently attempting to "Specialize" in a profession for the first time post 9.2.3 will cause a server-side crash that leads to the client disconnecting and the server incompletely freezing (still accepts connections but doesn't actually seem to update anything of  value). Additionally when attempting to report this crash ingame it allows me to enter the details then asks "Open This Page in Your Browser?" I click "yes" nothing opens...

Logs:
[log_210201043710.log](https://github.com/StrangeLoopGames/EcoIssues/files/5908124/log_210201043710.log)
[log_210201082055.log](https://github.com/StrangeLoopGames/EcoIssues/files/5908125/log_210201082055.log)
[log_210201083053.log](https://github.com/StrangeLoopGames/EcoIssues/files/5908126/log_210201083053.log)
[log_210201084609.log](https://github.com/StrangeLoopGames/EcoIssues/files/5908127/log_210201084609.log)
[log_210201043045.log](https://github.com/StrangeLoopGames/EcoIssues/files/5908128/log_210201043045.log)

I didn't find an "Eco" Folder in App Data at all let alone LocalLow so I'm assuming it's not linked to the player as no game dir exists let alone a log.

**Comment by MystLeissa** [comment]
[Test World.zip](https://github.com/StrangeLoopGames/EcoIssues/files/5908138/Test.World.zip)

These are my server save files...in case you want to test with them. I had to rip them out because I restored a much earlier backup but they are post-bug occurance. Some of the laws were removed from the live save in order to attempt to debug the crash however, this is my backup dir:




[Test World-2021-02-01-20.07.43.zip](https://github.com/StrangeLoopGames/EcoIssues/files/5908150/Test.World-2021-02-01-20.07.43.zip)

**Comment by MystLeissa** [comment]
Server encountered an exception:

<size=60.00%>Exception: ArgumentNullException
Message:Value cannot be null. (Parameter 'source')
Source:System.Linq

System.ArgumentNullException: Value cannot be null. (Parameter 'source')
   at System.Linq.ThrowHelper.ThrowArgumentNullException(ExceptionArgument argument)
   at System.Linq.Enumerable.Count[TSource](IEnumerable`1 source)
   at Eco.Gameplay.Civics.Pay_LegalAction.GetAndCheckValues(IContextObject context)
   at Eco.Gameplay.Civics.MoneyTransfer_LegalAction.Do(LocString description, IContextObject action, Law law)
   at Eco.Gameplay.Civics.MoneyTransfer_LegalAction.Perform(Law law, GameAction action)
   at Eco.Gameplay.Civics.LegalAction.PerformAndProcessResullt(Law law, GameAction action)
   at Eco.Gameplay.Civics.IfThenBlock.TryPerform(Law law, GameAction action, LawSection section, Int32 index, StringBuilder sbDebug)
   at Eco.Gameplay.Civics.Laws.LawSection.CheckConditionsAndApplyEffects(Law law, GameAction action)
   at Eco.Gameplay.Civics.Laws.LawSection.Perform(Law law, GameAction action)
   at Eco.Gameplay.Civics.Laws.Law.Perform(GameAction action)
   at Eco.Gameplay.Civics.Laws.LawManager.Perform(GameAction action)
   at Eco.Gameplay.GameActions.GameActionManager.TryPerformActions(User user, PackFlags flags, List`1 actions, IEnumerable`1 preTests, IEnumerable`1 postEffects)
   at Eco.Gameplay.GameActions.GameActionManager.TryPerformActions(User user, GameActionPack actionPack, Boolean notify)
   at Eco.Gameplay.GameActions.GameActionManager.TryPerform(GameActionPack pack, Boolean notify)
   at Eco.Gameplay.Skills.Skill.AddPoint(Player player, Boolean spendSkillPoints)
   at Eco.Gameplay.Skills.Skill.AddPoint(Player player)
Outer Exceptions:
Exception has been thrown by the target of an invocation.
Exception has been thrown by the target of an invocation.</size>

Log File:
Initialize engine version: 2019.4.15f1 (fbf367ac14e9)
[Subsystems] Discovering subsystems at path D:/SteamLibrary/steamapps/common/Eco/Eco_Data/UnitySubsystems
GfxDevice: creating device client; threaded=1
Direct3D:
    Version:  Direct3D 11.0 [level 11.1]
    Renderer: AMD Radeon RX 5700 XT (ID=0x731f)
    Vendor:   
    VRAM:     8152 MB
    Driver:   27.20.14501.28009
D3D11 device created for Microsoft Media Foundation video decoding.
<RI> Initializing input.

<RI> Input initialized.

<RI> Initialized touch support.

UnloadTime: 0.430300 ms
DontDestroyOnLoad only works for root GameObjects or components on root GameObjects. 
(Filename:  Line: 589)

WwiseUnity: Wwise(R) SDK Version 2019.2.5 Build 7349.
UnityEngine.Logger:LogFormat(LogType, String, Object[])
UnityEngine.Debug:LogFormat(String, Object[])
AkWwiseInitializationSettings:InitializeSoundEngine()
AkInitializer:OnEnable()
 
(Filename: C:\buildslave\unity\build\Runtime/Export/Debug/Debug.bindings.h Line: 35)

WwiseUnity: Setting Plugin DLL path to: D:/SteamLibrary/steamapps/common/Eco/Eco_Data\Plugins\x86_64
UnityEngine.Logger:Log(LogType, Object)
UnityEngine.Debug:Log(Object)
AkCommonPlatformSettings:get_AkInitializationSettings()
AkWwiseInitializationSettings:InitializeSoundEngine()
AkInitializer:OnEnable()
 
(Filename: C:\buildslave\unity\build\Runtime/Export/Debug/Debug.bindings.h Line: 35)

WwiseUnity: Sound engine initialized successfully.
UnityEngine.Logger:Log(LogType, Object)
UnityEngine.Debug:Log(Object)
AkWwiseInitializationSettings:InitializeSoundEngine()
AkInitializer:OnEnable()
 
(Filename: C:\buildslave\unity\build\Runtime/Export/Debug/Debug.bindings.h Line: 35)

The referenced script on this Behaviour (Game Object 'Camera') is missing! 
(Filename:  Line: 334)

The referenced script on this Behaviour (Game Object 'Table') is missing! 
(Filename:  Line: 334)

DontDestroyOnLoad only works for root GameObjects or components on root GameObjects.
SteamManager:Awake()

[ line 589] 
(Filename:  Line: 589)

Ambiguous UI component types ignored: SeriesPopupUI,BaseUI,ViewEditorUI,AutoViewUI
UnityEngine.Logger:Log(LogType, Object)
UnityEngine.Debug:Log(Object)
UI.UIManager:Start()
 
(Filename: C:\buildslave\unity\build\Runtime/Export/Debug/Debug.bindings.h Line: 35)

[09:17:04.432] Connecting to 192.168.1.19:3000...

EcoEngine.Logging.LogManager:LogFormat(LogType, Object, String, Object[])
UnityEngine.Logger:Log(LogType, Object)
UnityEngine.Debug:Log(Object)
Eco.Shared.Utils.Log:WriteLine(LocString)
Eco.Shared.Networking.NetworkClient:ConnectInternal(IPEndPoint)
Eco.Shared.Networking.NetworkClient:ConnectRoundRobin()
Eco.Shared.Networking.NetworkClient:Connect(IPEndPoint)
Eco.Shared.Networking.NetworkClient:Connect(String, Int32)
Eco.Shared.Networking.NetworkClient:Connect(ServerInfo)
Steamworks.RulesFailedToRespond:Invoke()
UnityEngine.Events.InvokableCall:Invoke()
UnityEngine.Events.UnityEvent:Invoke()
System.UnhandledExceptionEventHandler:Invoke(Object, UnhandledExceptionEventArgs)
UnityEngine.EventSystems.ExecuteEvents:Execute(GameObject, BaseEventData, EventFunction`1)
UnityEngine.EventSystems.StandaloneInputModule:ReleaseMouse(PointerEventData, GameObject)
UnityEngine.EventSystems.StandaloneInputModule:ProcessMousePress(MouseButtonEventData)
UnityEngine.EventSystems.StandaloneInputModule:ProcessMouseEvent(Int32)
UnityEngine.EventSystems.StandaloneInputModule:Process()
UnityEngine.EventSystems.EventSystem:Update()
 
(Filename: C:\buildslave\unity\build\Runtime/Export/Debug/Debug.bindings.h Line: 35)

The referenced script on this Behaviour (Game Object 'LayerViewer') is missing! 
(Filename:  Line: 334)

Unable to open archive file: D:/SteamLibrary/steamapps/common/Eco/Eco_Data/StreamingAssets/desc.unity3d 
(Filename:  Line: 584)

Failed to read data for the AssetBundle 'StreamingAssets\desc.unity3d'. 
(Filename:  Line: 449)

[09:17:10.431] Missing sprite in MiscIconMap.
EcoEngine.Logging.LogManager:LogFormat(LogType, Object, String, Object[])
UnityEngine.Logger:Log(LogType, Object)
UnityEngine.Debug:LogError(Object)
ModBundleManager:SetupSprites()
<ReceiveModData>d__47:MoveNext()
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)
 
(Filename: C:\buildslave\unity\build\Runtime/Export/Debug/Debug.bindings.h Line: 35)

[09:17:10.676] Missing following objects: AirPollutionGeneratorObject, CountryMarkerObject, FedeartionMarkerObject, HomesteadMarkerObject, TownMarkerObject
EcoEngine.Logging.LogManager:LogFormat(LogType, Object, String, Object[])
UnityEngine.Logger:Log(LogType, Object)
UnityEngine.Debug:LogWarning(Object)
<BuildFromModKit>d__8:MoveNext()
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)
<ReceiveModData>d__47:MoveNext()
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)
 
(Filename: C:\buildslave\unity\build\Runtime/Export/Debug/Debug.bindings.h Line: 35)

[09:17:14.559] ChatLogEntryNoFade doesn't support pooling. It may cause problems in runtime. Either add [SupportsPooling] attribute to all it's and it's children components or implement pool supporting interface(s) (IPoolRentAware, IPoolReturnAware). For third-party components you can use PoolingExtensions.RegisterPoolableType. List of components missing pooling support: ChatLogEntryNoFade (ChatLogEntry)
EcoEngine.Logging.LogManager:LogFormat(LogType, Object, String, Object[])
UnityEngine.Logger:Log(LogType, Object)
UnityEngine.Debug:LogWarning(Object)
EcoEngine.Pooling.ObjectPool:SetPrefab(PooledObject)
EcoEngine.Pooling.PooledObjectOwnerExtensions:PoolInstantiate(GameObject, Transform, Boolean, IPooledObjectOwner)
Assets.UI.Scripts.ChatLogUI:CreateAtIndex(Int32, Transform)
Assets.UI.Scripts.Utilities.CreateAtIndexFunc:Invoke(Int32, Transform)
Assets.UI.Scripts.Utilities.<>c__DisplayClass29_0:<MakeObject>b__0()
UnityEngine.AndroidJavaRunnable:Invoke()
Assets.UI.Scripts.Utilities.SparseScrollRect:LateUpdate()
 
(Filename: C:\buildslave\unity\build\Runtime/Export/Debug/Debug.bindings.h Line: 35)

[09:17:21.454] Missing Connecting
EcoEngine.Logging.LogManager:LogFormat(LogType, Object, String, Object[])
UnityEngine.Logger:Log(LogType, Object)
UnityEngine.Debug:LogError(Object)
UI.ConnectionUI:GotoState(ConnectState)
UI.ConnectionUI:ManualUpdate()
UI.UIManager:Update()
 
(Filename: C:\buildslave\unity\build\Runtime/Export/Debug/Debug.bindings.h Line: 35)

[09:17:38.082] Disconnected with error: Server encountered an exception:

<size=60.00%>Exception: ArgumentNullException
Message:Value cannot be null. (Parameter 'source')
Source:System.Linq

System.ArgumentNullException: Value cannot be null. (Parameter 'source')
   at System.Linq.ThrowHelper.ThrowArgumentNullException(ExceptionArgument argument)
   at System.Linq.Enumerable.Count[TSource](IEnumerable`1 source)
   at Eco.Gameplay.Civics.Pay_LegalAction.GetAndCheckValues(IContextObject context)
   at Eco.Gameplay.Civics.MoneyTransfer_LegalAction.Do(LocString description, IContextObject action, Law law)
   at Eco.Gameplay.Civics.MoneyTransfer_LegalAction.Perform(Law law, GameAction action)
   at Eco.Gameplay.Civics.LegalAction.PerformAndProcessResullt(Law law, GameAction action)
   at Eco.Gameplay.Civics.IfThenBlock.TryPerform(Law law, GameAction action, LawSection section, Int32 index, StringBuilder sbDebug)
   at Eco.Gameplay.Civics.Laws.LawSection.CheckConditionsAndApplyEffects(Law law, GameAction action)
   at Eco.Gameplay.Civics.Laws.LawSection.Perform(Law law, GameAction action)
   at Eco.Gameplay.Civics.Laws.Law.Perform(GameAction action)
   at Eco.Gameplay.Civics.Laws.LawManager.Perform(GameAction action)
   at Eco.Gameplay.GameActions.GameActionManager.TryPerformActions(User user, PackFlags flags, List`1 actions, IEnumerable`1 preTests, IEnumerable`1 postEffects)
   at Eco.Gameplay.GameActions.GameActionManager.TryPerformActions(User user, GameActionPack actionPack, Boolean notify)
   at Eco.Gameplay.GameActions.GameActionManager.TryPerform(GameActionPack pack, Boolean notify)
   at Eco.Gameplay.Skills.Skill.AddPoint(Player player, Boolean spendSkillPoints)
   at Eco.Gameplay.Skills.Skill.AddPoint(Player player)
Outer Exceptions:
Exception has been thrown by the target of an invocation.
Exception has been thrown by the target of an invocation.</size>
EcoEngine.Logging.LogManager:LogFormat(LogType, Object, String, Object[])
UnityEngine.Logger:Log(LogType, Object)
UnityEngine.Debug:LogWarning(Object)
NetworkManager:OnClientDisconnected(String)
System.Action`1:Invoke(T)
Eco.Shared.Networking.NetworkClient:HandleDisconnect(String)
UnityEngine.Experimental.Audio.SampleFramesHandler:Invoke(AudioSampleProvider, UInt32)
Eco.Networking.ENet.ENetUdpClient:OnDisconnected(ENetConnection, DisconnectReason)
Eco.Networking.ENet.ENetUdpPeer:HandleEvent()
Eco.Networking.ENet.ENetUdpPeer:Service(Int32)
<>c__DisplayClass63_0:<Initialize>b__0()
Steamworks.RulesFailedToRespond:Invoke()
System.Action`1:Invoke(T)
System.Threading.ExecutionContext:RunInternal(ExecutionContext, ContextCallback, Object, Boolean)
System.Threading.ExecutionContext:Run(ExecutionContext, ContextCallback, Object)
Steamworks.RulesFailedToRespond:Invoke()
 
(Filename: C:\buildslave\unity\build\Runtime/Export/Debug/Debug.bindings.h Line: 35)

[09:17:38.422] Destroyed FoodStatusUI has active subscriptions for: CachedHouseValueView.HousingSkillRate

EcoEngine.Logging.LogManager:LogFormat(LogType, Object, String, Object[])
UnityEngine.Logger:Log(LogType, Object)
UnityEngine.Debug:LogError(Object)
Eco.Shared.Utils.Log:WriteErrorLineLoc(FormattableString)
GenericClient`1:ReportNonReleasedSubscriptions()
GenericClient`1:<OnDisconnect>b__38_0()
UnityEngine.AndroidJavaRunnable:Invoke()
<InOneFrame>d__16:MoveNext()
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)
 
(Filename: C:\buildslave\unity\build\Runtime/Export/Debug/Debug.bindings.h Line: 35)

[09:17:38.422] Destroyed ViewUIContainer has active subscriptions for: UserTaskManagerView.Tasks

EcoEngine.Logging.LogManager:LogFormat(LogType, Object, String, Object[])
UnityEngine.Logger:Log(LogType, Object)
UnityEngine.Debug:LogError(Object)
Eco.Shared.Utils.Log:WriteErrorLineLoc(FormattableString)
GenericClient`1:ReportNonReleasedSubscriptions()
GenericClient`1:<OnDisconnect>b__38_0()
UnityEngine.AndroidJavaRunnable:Invoke()
<InOneFrame>d__16:MoveNext()
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)
 
(Filename: C:\buildslave\unity\build\Runtime/Export/Debug/Debug.bindings.h Line: 35)

[09:17:38.422] Destroyed UserTaskManagerUI has active subscriptions for: UserTaskManagerView.Tasks

EcoEngine.Logging.LogManager:LogFormat(LogType, Object, String, Object[])
UnityEngine.Logger:Log(LogType, Object)
UnityEngine.Debug:LogError(Object)
Eco.Shared.Utils.Log:WriteErrorLineLoc(FormattableString)
GenericClient`1:ReportNonReleasedSubscriptions()
GenericClient`1:<OnDisconnect>b__38_0()
UnityEngine.AndroidJavaRunnable:Invoke()
<InOneFrame>d__16:MoveNext()
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)
 
(Filename: C:\buildslave\unity\build\Runtime/Export/Debug/Debug.bindings.h Line: 35)

Unloading 9 Unused Serialized files (Serialized files now loaded: 0)

Unloading 8594 unused Assets to reduce memory usage. Loaded Objects now: 118188.
Total: 304.469300 ms (FindLiveObjects: 11.259400 ms CreateObjectMapping: 17.820900 ms MarkObjects: 266.381100 ms  DeleteObjects: 9.007600 ms)


May be more helpful than the server logs...

**Comment by MystLeissa** [comment]
Ok I managed to restore a backup and fix it somewhat, apparently it doesn't like the "Profession Refund" Law that was contained within my save data. It was a law to provide a small amount of "admin" currency (non-material minted currency) back to a player for gaining a profession...when I removed that law it let me spec right.

**Comment by SlayksWood** [comment]
Step to reproduce:
- download this server:
> Test World-2021-02-01-20.07.43.zip
- try to take any specialty.
- Crash.

**Comment by SlayksWood** [comment]
Also can reproduce in fresh world. 
- make law like this:
![image](https://user-images.githubusercontent.com/45708377/106579099-58b5d500-6551-11eb-9251-62a51ee015c8.png)
- win election and take specialty.

**Comment by SlayksWood** [comment]
Seems this happens for all laws where you select 'Prevent if Unable to play' = yes. I create this law:
![image](https://user-images.githubusercontent.com/45708377/106581660-278ad400-6554-11eb-8eaf-659fd7549c4e.png)
and when I tried to place Workbench I was crashed:
```
<size=60,00%>Exception: ArgumentNullException
Message:Value cannot be null. (Parameter 'source')
Source:System.Linq

System.ArgumentNullException: Value cannot be null. (Parameter 'source')
   at System.Linq.ThrowHelper.ThrowArgumentNullException(ExceptionArgument argument)
   at System.Linq.Enumerable.Count[TSource](IEnumerable`1 source)
   at Eco.Gameplay.Civics.Pay_LegalAction.GetAndCheckValues(IContextObject context)
   at Eco.Gameplay.Civics.MoneyTransfer_LegalAction.Do(LocString description, IContextObject action, Law law)
   at Eco.Gameplay.Civics.MoneyTransfer_LegalAction.Perform(Law law, GameAction action)
   at Eco.Gameplay.Civics.LegalAction.PerformAndProcessResullt(Law law, GameAction action)
   at Eco.Gameplay.Civics.IfThenBlock.TryPerform(Law law, GameAction action, LawSection section, Int32 index, StringBuilder sbDebug)
   at Eco.Gameplay.Civics.Laws.LawSection.CheckConditionsAndApplyEffects(Law law, GameAction action)
   at Eco.Gameplay.Civics.Laws.LawSection.Perform(Law law, GameAction action)
   at Eco.Gameplay.Civics.Laws.Law.Perform(GameAction action)
   at Eco.Gameplay.Civics.Laws.LawManager.Perform(GameAction action)
   at Eco.Gameplay.GameActions.GameActionManager.TryPerformActions(User user, PackFlags flags, List`1 actions, IEnumerable`1 preTests, IEnumerable`1 postEffects)
   at Eco.Gameplay.GameActions.GameActionManager.TryPerformActions(User user, GameActionPack actionPack, Boolean notify)
   at Eco.Gameplay.GameActions.GameActionManager.TryPerform(GameActionPack pack, Boolean notify)
   at Eco.Gameplay.Objects.WorldObjectManager.TryPlaceWorldObject(Player player, WorldObjectItem item, Vector3i pos, Quaternion rot)
   at Eco.Gameplay.Items.WorldObjectItem.<>c__DisplayClass26_1.<OnActRight>b__1()
   at Eco.Gameplay.Items.WorldObjectItem.TryPlaceObject(Player player, Vector3i position, Quaternion rotation, Action successCallback)
   at Eco.Gameplay.Items.WorldObjectItem.<>c__DisplayClass26_0.<OnActRight>b__0(BSONObject response)
   at Eco.Shared.Networking.RPCManager.<>c__DisplayClass22_0`1.<Call>b__0(Object o)
   at Eco.Shared.Networking.RPCManager.HandleQueryResponse(BSONObject bson)
   at Eco.Plugins.Networking.NetworkServer.Eco.Shared.Networking.INetworkEventHandler.ReceiveEvent(INetClient client, NetworkEvent netEvent, BSONObject bsonObj)
   at Eco.Plugins.Networking.Client.<.ctor>b__70_3()</size>
EcoEngine.Logging.LogManager:LogFormat(LogType, Object, String, Object[])
```

**Comment by SlayksWood** [comment]
@MystLeissa Thank you a lot for helping.
If you have a SLG account and want a little reward, send an e-mail to dennis@strangeloopgames.com with your SLG ID and link with this issue.

**Comment by WeaselDog** [comment]
This is fixed in 9.2.4
